src/main/java/de/tum/in/www1/artemis/service/exam/ExamRegistrationService.java

```diff
@@ -74,30 +81,53 @@ public ExamRegistrationService(ExamRepository examRepository, UserService userSe
      * This method first tries to find the student in the internal Artemis user database (because the user is most probably already using Artemis).
      * In case the user cannot be found, we additionally search the (TUM) LDAP in case it is configured properly.
      *
-     * @param courseId    the id of the course
-     * @param examId      the id of the exam
-     * @param studentDTOs the list of students (with at least registration number) who should get access to the exam
+     * @param courseId     the id of the course
+     * @param examId       the id of the exam
+     * @param examUserDTOs the list of students (with at least registration number) who should get access to the exam
      * @return the list of students who could not be registered for the exam, because they could NOT be found in the Artemis database and could NOT be found in the TUM LDAP
      */
-    public List<StudentDTO> registerStudentsForExam(Long courseId, Long examId, List<StudentDTO> studentDTOs) {
+    public List<ExamUserDTO> registerStudentsForExam(Long courseId, Long examId, List<ExamUserDTO> examUserDTOs) {
         var course = courseRepository.findByIdElseThrow(courseId);
-        var exam = examRepository.findWithRegisteredUsersById(examId).orElseThrow(() -> new EntityNotFoundException("Exam", examId));
+        var exam = examRepository.findWithExamUsersById(examId).orElseThrow(() -> new EntityNotFoundException("Exam", examId));
 
         if (exam.isTestExam()) {
             throw new AccessForbiddenException("Registration of students is only allowed for real exams");
         }
 
-        List<StudentDTO> notFoundStudentsDTOs = new ArrayList<>();
-        for (var studentDto : studentDTOs) {
-            var registrationNumber = studentDto.getRegistrationNumber();
-            var login = studentDto.getLogin();
-            var email = studentDto.getEmail();
+        List<ExamUserDTO> notFoundStudentsDTOs = new ArrayList<>();
+        for (var examUserDto : examUserDTOs) {
+            var registrationNumber = examUserDto.registrationNumber();
+            var login = examUserDto.login();
+            var email = examUserDto.email();
             Optional<User> optionalStudent = userService.findUserAndAddToCourse(registrationNumber, course.getStudentGroupName(), Role.STUDENT, login, email);
             if (optionalStudent.isEmpty()) {
-                notFoundStudentsDTOs.add(studentDto);
+                notFoundStudentsDTOs.add(examUserDto);
             }
             else {
-                exam.addRegisteredUser(optionalStudent.get());
+                ExamUser registeredExamUserCheck = examUserRepository.findByExamIdAndUserId(exam.getId(), optionalStudent.get().getId());
+                if (!exam.getExamUsers().contains(registeredExamUserCheck) && !optionalStudent.get().getAuthorities().contains(ADMIN_AUTHORITY)
+                        && !optionalStudent.get().getGroups().contains(course.getInstructorGroupName())) {
```

Comment:
We have the authCheckService for checking if a user has a specific role in a course or exercise

Reply:
Now i use the authCheck service :+1: